multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,67 +79,227 @@ func NewNodeReconciler(
-	// check if we can improve this with 'Owns' or other methods.
-	var gwIP, internalIP string
-	if isGW {
-		if internalIP, gwIP, err = r.getGatawayNodeIP(node); err != nil {
-			klog.ErrorS(err, "There is no valid Gateway IP for Node, will retry later when there is any new Node update", "node", node.Name)
+	if !nodeIsReady {
+		if isActiveGateway {
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
+				return ctrl.Result{}, err
+			}
 			return ctrl.Result{}, nil
 		}
+		if hasGWAnnotation {
+			r.gatewayCandidates.Insert(req.Name)
+		}
+		return ctrl.Result{}, nil
 	}
-	if err := r.Client.Get(ctx, gwNamespacedName, gw); err != nil {
-		if apierrors.IsNotFound(err) && isGW {
-			gw.GatewayIP = gwIP
-			gw.InternalIP = internalIP
-			if err := r.Client.Create(ctx, gw, &client.CreateOptions{}); err != nil {
+
+	var err error
+	var gwIP, internalIP string
+	// When Node is ready without Gateway annotation
+	if !hasGWAnnotation {
+		if isActiveGateway {
+			// The annotation is removed from the Node, we need recreate a new Gateway.
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
 				return ctrl.Result{}, err
 			}
-			return ctrl.Result{}, nil
 		}
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+		return ctrl.Result{}, nil
 	}
 
-	if !isGW {
-		err := r.Client.Delete(ctx, gw, &client.DeleteOptions{})
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+	// When the Node is ready with the Gateway annotation
+	r.gatewayCandidates.Insert(node.Name)
+	internalIP, gwIP, err = r.getGatawayNodeIP(node)
+	if err != nil {
+		klog.ErrorS(err, "There is no valid Gateway IP for Node", "node", node.Name)
+		if isActiveGateway {
+			if err := r.recreateActiveGateway(ctx, gw); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+		return ctrl.Result{}, nil
 	}
-
-	gw.GatewayIP = gwIP
 	gw.InternalIP = internalIP
-	if err := r.Client.Update(ctx, gw, &client.UpdateOptions{}); err != nil {
-		return ctrl.Result{}, err
+	gw.GatewayIP = gwIP
+	if isActiveGateway {
+		if err := r.updateActiveGateway(ctx, gw); err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+	if noActiveGateway {
+		if err := r.createGateway(gw); err != nil {
+			return ctrl.Result{}, err
+		}
 	}
 	return ctrl.Result{}, nil
 }
 
+// initializeActiveGateway initialize 'activeGateway' and 'gatewayCandidates'
+// during controller startup.
+func (r *NodeReconciler) initializeActiveGateway() error {
+	ctx := context.Background()
+	nodeList := &corev1.NodeList{}
+	if err := r.Client.List(ctx, nodeList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	for _, n := range nodeList.Items {
+		if _, isGW := n.Annotations[common.GatewayAnnotation]; isGW {
+			r.gatewayCandidates.Insert(n.Name)
+
+		}
+	}
+
+	gwList := &mcsv1alpha1.GatewayList{}
+	if err := r.Client.List(ctx, gwList, &client.ListOptions{}); err != nil {
+		return err
+	}
+	// Gateway webhook guarantees that there are at most one Gateway in the member cluster.
+	size := len(gwList.Items)
+	switch size {
```

Comment:
If feel the following code is hard to understand. I would suggest to handle only the case that a Gateway exists:
* If it is no longer a valid Gateway, remove it;
* If it is a valid Gateway, set it to activeGateway.

Then we just return, and let Reconcile() handle Gateway changes, so initializeActiveGateway() is only for initializing the reconciler state, and removing stale Gateway.

Reply:
Refined, thanks.